packages/core/contracts/financial-templates/common/financial-product-libraries/PreExpirationIdentifierTransformationFinancialProductLibrary.sol

```diff
@@ -0,0 +1,59 @@
+ * & if it is at or after expiration then the original identifier is returned. This library enables self referential
+ * TWAP identifier to be used on synthetics pre-expiration, in conjunction with a separate identifier at expiration.
+ */
+contract PreExpirationIdentifierTransformationFinancialProductLibrary is FinancialProductLibrary, Ownable {
+    mapping(address => bytes32) financialProductTransformedIdentifiers;
+
+    /**
+     * @notice Enables the deployer of the library to set the transformed identifier for an associated financial product.
+     * @param financialProduct address of the financial product.
+     * @param transformedIdentifier the identifier for the financial product to be used if the contract is post expiration.
+     * @dev Note: a) Only the owner (deployer) of this library can set identifier transformations b) The identifier Cant
+     * be set to blank. c) A transformed price can only be set once to prevent the deployer from changing it after the fact.
+     * d)  financialProduct must exposes an expirationTimestamp method.
+     */
+    function setFinancialProductTransformedIdentifier(address financialProduct, bytes32 transformedIdentifier)
+        public
+        onlyOwner
+    {
+        require(transformedIdentifier != "", "Cant set to empty transformation");
+        require(financialProductTransformedIdentifiers[financialProduct] == "", "Transformation already set");
+        require(ExpiringContractInterface(financialProduct).expirationTimestamp() != 0, "Invalid EMP contract");
+        financialProductTransformedIdentifiers[financialProduct] = transformedIdentifier;
+    }
+
+    /**
+     * @notice Returns the transformed identifier associated with a given financial product address.
+     * @param financialProduct address of the financial product.
+     * @return transformed identifier for the associated financial product.
+     */
+    function getTransformedIdentifierForFinancialProduct(address financialProduct) public view returns (bytes32) {
+        return financialProductTransformedIdentifiers[financialProduct];
+    }
+
+    /**
+     * @notice Returns a transformed price identifier if the contract is pre-expiration and no transformation if post.
+     * @param identifier input price identifier to be transformed.
+     * @param requestTime timestamp the identifier is to be used at.
+     * @return transformedPriceIdentifier the input price identifier with the transformation logic applied to it.
+     */
+    function transformPriceIdentifier(bytes32 identifier, uint256 requestTime) public view override returns (bytes32) {
+        require(financialProductTransformedIdentifiers[msg.sender] != "", "Caller has no transformation");
+        // If the request time is before contract expiration then return the transformed identifier. Else, return the
+        // transformed price identifier.
```

Comment:
Just curious -- why is the pre-expiration one the "transformed" one and the expiry one the "non-transformed" one?

Reply:
My thinking was that in most scenarios you're payout as a user will be the non-transformed (or the parameterized) identifier. The "transformed" in this scenario is used only in case of liquidation and so is rarer than the non-transformed. 

For example, in the case of uGas it settles to the non-transformed price which is the long-running median and in the case of liquidation(with is less common) only then the identifier is transformed. You can think of it almost like being able to set separate identifiers for liquidations to that used at settlement.

Reply:
Yeah, that makes sense.